### PR TITLE
US657247 : Full bundle support

### PR DESCRIPTION
**Pull request by RaghuReddy**
## Description
Full bundle support

## Links
Link Type   | Link
------      | ------
Rally Issue | [US657247]
Func Spec   | [Func Spec]

## Checklist
- [x] Pull Request Created
- [ ] Code Review Completed
- [ ] Veracode scan results addressed
- [ ] SonarQube scan results addressed
- [ ] TPSR has been submitted (if applicable) 
- [x] Unit tests have been added
- [ ] Integration/Functional test cases have been written and automated
- [ ] Func Spec has been updated as needed
- [ ] Rally Issue(s) are in an engineering completed state

### CODE REVIEW

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleEntityBuilder.java [hunk 1]
```diff
@@ -52,8 +52,15 @@
             List<Entity> entities = new ArrayList<>();
             entityBuilders.forEach(builder -> entities.addAll(builder.build(bundle, bundleType, document)));
             final Element fullBundle = bundleDocumentBuilder.build(document, entities);
+            BundleMetadata fullBundleMetadata = null;
```
Comment (uppoju) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleEntityBuilder.java:55]: is it not required to generate metadata for build-environment-bundle task ?
In annotated entity case, we are generating metadata for both the tasks

Comment (RaghuReddy) [reply config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleEntityBuilder.java:55]: It is not required, we should correct it for annotated entity case as well

Comment (RaghuReddy) [reply config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleEntityBuilder.java:55]: BTW, I do not think we are generating metadata for build-environment-bundle task (annotated case), as annotations are not ignored for env entities.

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/AnnotableEntity.java [hunk 2]
```diff
@@ -50,6 +50,7 @@
             AnnotatedEntity<GatewayEntity> annotatedEntity = new AnnotatedEntity(this);
             annotations.forEach(annotation -> {
                 if (ANNOTATION_TYPE_BUNDLE.equalsIgnoreCase(annotation.getType())) {
+                    annotatedEntity.setId(annotation.getId());
```
Comment (uppoju) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/AnnotableEntity.java:53]: currently we are using this id for entity, we should have separate field for bundle id

Comment (RaghuReddy) [reply config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/AnnotableEntity.java:53]: Yes correct, we are addressing that in other story, i will revert this change

File: gateway-developer-plugin/src/main/java/com/ca/apim/gateway/cagatewayconfig/BuildEnvironmentBundleTask.java [hunk 3]
```diff
@@ -52,7 +51,13 @@ DirectoryProperty getInto() {
         return into;
     }
 
+    @Input
```
Comment (uppoju) [inline gateway-developer-plugin/src/main/java/com/ca/apim/gateway/cagatewayconfig/BuildEnvironmentBundleTask.java:54]: i think this should be optional

File: gateway-developer-plugin/src/main/java/com/ca/apim/gateway/cagatewayconfig/BuildFullBundleTask.java [hunk 4]
```diff
@@ -60,12 +66,18 @@ DirectoryProperty getInto() {
         return into;
     }
 
+    @Input
```
Comment (uppoju) [inline gateway-developer-plugin/src/main/java/com/ca/apim/gateway/cagatewayconfig/BuildFullBundleTask.java:69]: This should be optional

File: gateway-developer-plugin/src/main/java/com/ca/apim/gateway/cagatewayconfig/BuildEnvironmentBundleTask.java (hunk 3 above)
Comment (RaghuReddy) [reply gateway-developer-plugin/src/main/java/com/ca/apim/gateway/cagatewayconfig/BuildEnvironmentBundleTask.java:54]: Done

File: gateway-developer-plugin/src/main/java/com/ca/apim/gateway/cagatewayconfig/BuildFullBundleTask.java (hunk 4 above)
Comment (RaghuReddy) [reply gateway-developer-plugin/src/main/java/com/ca/apim/gateway/cagatewayconfig/BuildFullBundleTask.java:69]: done

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleEntityBuilder.java [hunk 5]
```diff
@@ -52,8 +52,15 @@
...
+            if (EntityBuilder.BundleType.DEPLOYMENT.equals(bundleType)) {
```
Comment (nikhilahuja) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleEntityBuilder.java:56]: You can directly use `==` operator, it an enum.

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleMetadataBuilder.java [hunk 6]
```diff
@@ -9,42 +9,97 @@
...
+     * @param projectGroupName
+     * @param projectVersion
+     * @return bundle metadata
+     */
+    private BundleMetadata buildFullBundleMetadata (final List<Entity> entities, final String projectName,
+                                                    final String projectGroupName, final String projectVersion) {
+        BundleMetadata.Builder builder = new BundleMetadata.Builder(BUNDLE_TYPE_ALL, idGenerator.generate(), projectName,
+                projectGroupName, projectVersion);
+        builder.description(StringUtils.EMPTY);
+        builder.tags(Collections.EMPTY_LIST);
```
Comment (nikhilahuja) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleMetadataBuilder.java:84]: It is recommended to use `Collections.emptyList()`

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/Entity.java [hunk 7]
```diff
@@ -96,17 +96,17 @@ public String getName() {
                 return name;
             }
 
+            @JsonIgnore
```
Comment (nikhilahuja) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/Entity.java:99]: This will conflict with `definedEntities`. There we need `id` and `guid`.

Comment (RaghuReddy) [reply config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/Entity.java:99]: @nikhilahuja i have tested that, no conflict

File: gateway-developer-plugin/src/main/java/com/ca/apim/gateway/cagatewayconfig/BuildEnvironmentBundleTask.java [hunk 8]
```diff
@@ -52,29 +51,41 @@ DirectoryProperty getInto() {
...
             throw new MissingEnvironmentException("Metadata file does not exist.");
         }
-        metaDataFiles.stream().forEach(metaDataFile-> {
-            final Pair<String, Map<String, String>> bundleEnvironmentValues = environmentConfigurationUtils.parseBundleMetadata(metaDataFile, configFolder.getAsFile().get());
+
+        metaDataFiles.stream().forEach(metaDataFile -> {
+            final Pair<String, Map<String, String>> bundleEnvironmentValues = environmentConfigurationUtils.parseBundleMetadata(metaDataFile, configFolder.getAsFile().getOrNull());
             if (null != bundleEnvironmentValues) {
-                final String bundleFileName = bundleEnvironmentValues.getLeft() + "." + configName.get() + ".environment.bundle";
+                final String bundleFileName = bundleEnvironmentValues.getLeft() + "-" + configName.get() + "env.install.bundle";
```
Comment (nikhilahuja) [inline gateway-developer-plugin/src/main/java/com/ca/apim/gateway/cagatewayconfig/BuildEnvironmentBundleTask.java:83]: Can you revert this change please. It will be a conflict with #234

File: gateway-developer-plugin/src/main/java/com/ca/apim/gateway/cagatewayconfig/BuildFullBundleTask.java [hunk 9]
```diff
@@ -79,21 +93,25 @@ DirectoryProperty getConfigFolder() {
...
         if(metaDataFiles.isEmpty()) {
             throw new MissingEnvironmentException("Metadata file does not exist.");
         }
+
         metaDataFiles.stream().forEach(metaDataFile-> {
-            final Pair<String, Map<String, String>> bundleEnvironmentValues = environmentConfigurationUtils.parseBundleMetadata(metaDataFile, configFolder.getAsFile().get());
+            final Pair<String, Map<String, String>> bundleEnvironmentValues = environmentConfigurationUtils.parseBundleMetadata(metaDataFile, configFolder.getAsFile().getOrNull());
             if (null != bundleEnvironmentValues) {
-                final String bundleFileName = bundleEnvironmentValues.getLeft() + "." + configName.get() + ".full.bundle";
+                final String bundleFileName = bundleEnvironmentValues.getLeft() + "-full.install.bundle";
```
Comment (nikhilahuja) [inline gateway-developer-plugin/src/main/java/com/ca/apim/gateway/cagatewayconfig/BuildFullBundleTask.java:104]: Please revert this change. It will be a conflict with #234

File: gateway-developer-plugin/src/main/java/com/ca/apim/gateway/cagatewayconfig/CAGatewayDeveloper.java [hunk 10]
```diff
@@ -111,10 +116,14 @@ private static BuildEnvironmentBundleTask createBuildEnvironmentBundleTask(@NotN
...
         // Create build-full-bundle task
         final BuildFullBundleTask buildFullBundleTask = project.getTasks().create(BUILD_FULL_BUNDLE, BuildFullBundleTask.class, t -> {
+            t.getEnvironmentConfig().set(pluginConfig.getEnvironmentConfig());
             t.getDependencyBundles().setFrom(project.getConfigurations().getByName(BUNDLE_CONFIGURATION));
             t.getDetemplatizeDeploymentBundles().set(pluginConfig.getDetemplatizeDeploymentBundles().getOrElse(true));
             t.getInto().set(pluginConfig.getBuiltEnvironmentBundleDir());
-            t.getConfigFolder().set(pluginConfig.getConfigFolder());
+            t.getConfigFolder().set(new DefaultProvider<>(() -> {
+                Directory dir = pluginConfig.getConfigFolder().get();
+                return dir.getAsFile().exists() ? dir : null;
```
Comment (nikhilahuja) [inline gateway-developer-plugin/src/main/java/com/ca/apim/gateway/cagatewayconfig/CAGatewayDeveloper.java:125]: `dir` could be `null`. Have a null check in here

File: gateway-developer-plugin/src/main/java/com/ca/apim/gateway/cagatewayconfig/CAGatewayDeveloper.java [hunk 11]
```diff
@@ -101,7 +102,11 @@ private static BuildEnvironmentBundleTask createBuildEnvironmentBundleTask(@NotN
         // Create build-environment-bundle task
         final BuildEnvironmentBundleTask buildEnvironmentBundleTask = project.getTasks().create(BUILD_ENVIRONMENT_BUNDLE, BuildEnvironmentBundleTask.class, t -> {
             t.getInto().set(pluginConfig.getBuiltEnvironmentBundleDir());
-            t.getConfigFolder().set(pluginConfig.getConfigFolder());
+            t.getEnvironmentConfig().set(pluginConfig.getEnvironmentConfig());
+            t.getConfigFolder().set(new DefaultProvider<>(() -> {
+                Directory dir = pluginConfig.getConfigFolder().get();
+                return dir.getAsFile().exists() ? dir : null;
```
Comment (nikhilahuja) [inline gateway-developer-plugin/src/main/java/com/ca/apim/gateway/cagatewayconfig/CAGatewayDeveloper.java:108]: Have null check here as well.

Comment (RaghuReddy) [reply gateway-developer-plugin/src/main/java/com/ca/apim/gateway/cagatewayconfig/CAGatewayDeveloper.java:108]: @nikhilahuja , added a null check at caller side

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BuilderConstants.java [hunk 12]
```diff
@@ -29,6 +30,13 @@ public boolean test(Entity entity) {
         }
     };
 
+    public static final Predicate<Entity> FILTER_METADATA_NON_ENV_ENTITIES = new Predicate<Entity>() {
```
Comment (nikhilahuja) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BuilderConstants.java:33]: You should change name to something like `FILTER_NON_ENV_ENTITIES_EXCLUDING_FOLDER`

File: gateway-developer-plugin/src/main/java/com/ca/apim/gateway/cagatewayconfig/BuildFullBundleTask.java (hunk 9 above)
Comment (RaghuReddy) [reply gateway-developer-plugin/src/main/java/com/ca/apim/gateway/cagatewayconfig/BuildFullBundleTask.java:104]: @nikhilahuja , I will take changes from your PR

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleMetadataBuilder.java [hunk 13]
```diff
@@ -9,42 +9,97 @@
...
+     * @param annotatedBundle
+     * @param annotatedEntity
+     * @param dependentEntities
+     * @param projectName
+     * @param projectGroupName
+     * @param projectVersion
+     * @return bundle metadata
+     */
     public BundleMetadata build(final AnnotatedBundle annotatedBundle,
                                 final AnnotatedEntity<? extends GatewayEntity> annotatedEntity,
```
Comment (nikhilahuja) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleMetadataBuilder.java:46]: Can you remove this argument? `AnnotatedBundle` contains the annotatedEntity so we can use that.

File: gateway-developer-plugin/src/main/java/com/ca/apim/gateway/cagatewayconfig/BuildEnvironmentBundleTask.java (hunk 8 above)
Comment (RaghuReddy) [reply gateway-developer-plugin/src/main/java/com/ca/apim/gateway/cagatewayconfig/BuildEnvironmentBundleTask.java:83]: @nikhilahuja , do not worry about this, I will merge your PR once it is approved
